maths/bbp_formula.py

```diff
@@ -0,0 +1,105 @@
+def bbp_extract_pi_hex_digit(digit_position: int, precision: int = 1000) -> str:
+    """
+    Implement a popular pi-digit-extraction algorithm known as the 
+    Bailey-Borwein-Plouffe (BBP) formula to calulate the nth hex digit of pi.
+    Wikipedia page:
+    https://en.wikipedia.org/wiki/Bailey%E2%80%93Borwein%E2%80%93Plouffe_formula
+    @param digit_position: a positive integer representing the position of the digit to extract. 
+    The digit immediately after the decimal point is located at position 1.
+    @param precision: number of terms in the second summation to calculate.
+    A higher number reduces the chance of an error but increases the runtime.
+    @return: a hexadecimal digit representing the digit at the nth position
+    in pi's decimal expansion.
+    
+    >>> bbp_extract_pi_hex_digit(-10)
+    Traceback (most recent call last):
+      ...
+    ValueError: Digit position must be a positive integer
+    >>> bbp_extract_pi_hex_digit(1.7)
+    Traceback (most recent call last):
+      ...
+    ValueError: Digit position must be a positive integer
+    >>> bbp_extract_pi_hex_digit(0)
+    Traceback (most recent call last):
+      ...
+    ValueError: Digit position must be a positive integer
+    >>> bbp_extract_pi_hex_digit(2, -10)
+    Traceback (most recent call last):
+      ...
+    ValueError: Precision must be a nonnegative integer
+    >>> bbp_extract_pi_hex_digit(2, 1.6)
+    Traceback (most recent call last):
+      ...
+    ValueError: Precision must be a nonnegative integer
```

Comment:
Please put the passing tests first and the failing tests last.

Reply:
Are you talking about the doctests failing? If so I'm pretty sure they all pass because when I run the code on my local machine there are no "failed test" messages

Reply:
I think he means the tests that if you were to call that function with those parameters yourself you will get an error. IE `bbp_extract_pi_hex_digit(2, 1.6)` is a failing test because it calling the function with invalid parameters.